components/core/src/clp/ffi/Value.hpp

```diff
@@ -0,0 +1,224 @@
+ */
+template <typename Type>
+concept PrimitiveValueType = cIsValidPrimitiveValueType<Type>;
+
+/**
+ * Concept that defines primitive value types that are C++ fundamental types.
+ */
+template <typename Type>
+concept FundamentalValueType = cIsValidPrimitiveValueType<Type> && std::is_fundamental_v<Type>;
+
+/**
+ * Concept that defines move-constructable primitive value types.
+ */
+template <typename Type>
+concept MoveConstructableValueType
+        = cIsValidPrimitiveValueType<Type> && std::is_move_constructible_v<Type>
+          && (false == std::is_fundamental_v<Type>);
+
+/**
+ * Template struct that converts a given type into an immutable view type. By default, the immutable
+ * view type is the given type itself, meaning that the immutable view is a copy. Specialization is
+ * needed when the immutable view type is a const reference or some other types.
+ * @tparam Type
+ */
+template <typename Type>
+struct ImmutableViewTypeConverter {
+    using type = Type;
+};
+
+/**
+ * Specializes `std::string`'s immutable view type to `std::string_view`.
+ */
+template <>
+struct ImmutableViewTypeConverter<std::string> {
+    using type = std::string_view;
+};
+
+/**
+ * Specializes `clp::ir::EightByteEncodedTextAst`'s immutable view type its const reference.
+ */
+template <>
+struct ImmutableViewTypeConverter<clp::ir::EightByteEncodedTextAst> {
+    using type = clp::ir::EightByteEncodedTextAst const&;
+};
+
+/**
+ * Specializes `clp::ir::FourByteEncodedTextAst`'s immutable view type its const reference.
+ */
+template <>
+struct ImmutableViewTypeConverter<clp::ir::FourByteEncodedTextAst> {
+    using type = clp::ir::FourByteEncodedTextAst const&;
+};
+
+/**
+ * Template alias to the underlying typename of `ImmutableViewTypeConverter`.
+ * @tparam Type
+ */
+template <typename Type>
+using ImmutableViewType = typename ImmutableViewTypeConverter<Type>::type;
+
+class Value {
+public:
+    // Types
+    class OperationFailed : public TraceableException {
+    public:
+        OperationFailed(
+                ErrorCode error_code,
+                char const* const filename,
+                int line_number,
+                std::string message
+        )
+                : TraceableException{error_code, filename, line_number},
+                  m_message{std::move(message)} {}
+
+        [[nodiscard]] auto what() const noexcept -> char const* override {
+            return m_message.c_str();
+        }
+
+    private:
+        std::string m_message;
+    };
+
+    // Constructors
+    Value() = default;
+
+    /**
+     * Move constructs from the given rvalue.
+     * @tparam PrimitiveValue The type of the rvalue, which must be a move constructable
+     * non-reference type.
+     * @value
+     */
+    template <MoveConstructableValueType PrimitiveValue>
+    requires(false == std::is_reference_v<PrimitiveValue>)
+    Value(PrimitiveValue&& value) : m_value{std::forward<PrimitiveValue>(value)} {}
```

Comment:
The `requires` constraint here doesn't do anything since it's operating on `PrimitiveValue` which won't be a reference after the compiler performs reference collapsing (and either way, I think you intended for it to be `is_lvalue_reference_v`?). What I found is that to prevent callers from passing lvalue references, we need to add an assertion to the body:

`static_assert(std::is_rvalue_reference_v<decltype(value)>);`

If the caller passes in an lvalue or anything that collapses to an lvalue, this assertion will fail.